src/database.rs

```diff
@@ -76,41 +98,90 @@ pub fn get_data_from_db(
             .collect();
 
         if missing_sgs.is_empty() {
-            println!("No missing state groups");
+            // println!("No missing state groups");
             break;
         }
 
         missing_sgs.sort_unstable();
         missing_sgs.dedup();
 
-        println!("Missing {} state groups", missing_sgs.len());
+        // println!("Missing {} state groups", missing_sgs.len());
 
-        let map = get_missing_from_db(&mut client, &missing_sgs);
-        state_group_map.extend(map.into_iter());
+        // find state groups not picked up already and add them to the map
+        let map = get_missing_from_db(&mut client, &missing_sgs, min_state_group, max_group_found);
+        for (k, v) in map.into_iter() {
+            state_group_map.entry(k).or_insert(v);
+        }
     }
 
-    state_group_map
+    (state_group_map, max_group_found)
+}
+
+/// Returns the group ID of the last group to be compressed
+///
+/// This can be saved so that future runs of the compressor only
+/// continue from after this point
+///
+/// # Arguments
+///
+/// * `client`              -   A Postgres client to make requests with
+/// * `room_id`             -   The ID of the room in the database
+/// * `min_state_group`     -   The lower limit (non inclusive) of group id's to compress
+/// * 'groups_to_compress'  -   How many groups to compress
+/// * `max_state_group`     -   The upper bound on what this method can return
+fn find_max_group(
+    client: &mut Client,
+    room_id: &str,
+    min_state_group: Option<i64>,
+    groups_to_compress: Option<i64>,
+    max_state_group: Option<i64>,
+) -> i64 {
+    // Get list of state_id's in a certain room
+    let mut sql = "SELECT id FROM state_groups WHERE room_id = $1".to_string();
+
+    if let Some(max) = max_state_group {
+        sql = format!("{} AND id <= {}", sql, max)
+    }
+
+    // Adds additional constraint if a groups_to_compress has been specified
+    // Then sends query to the datatbase
+    let rows = if let (Some(min), Some(count)) = (min_state_group, groups_to_compress) {
+        let params: Vec<&dyn ToSql> = vec![&room_id, &min, &count];
+        client.query_raw(format!(r"{} AND id > $2 LIMIT $3", sql).as_str(), params)
```

Comment:
```suggestion
        client.query_raw(format!(r"{} AND id > $2 OFFSET $3 LIMIT 1", sql).as_str(), params)
```

I think works?

Reply:
I think it should be "OFFSET [count-1] LIMIT 1" but yes that works

Reply:
Hmmm I'm now getting an annoying error from postgres: `WrongType { postgres: Int4, rust: "i64" }`  which seems to be solved by changing groups_to_compress to an Option\<i32\>

Reply:
I see no particular reason not to just make it i32 though (because if you want to compress more than the 32 bit limit in one go then that's quite a special case! - and would be better done with the saving and restoring stuff I add in later)

Reply:
Hmm There's an issue with OFFSET preventing you from compressing the last chunk of a room

Previously, if there were 400 groups between min_state_group and the highest in the table and you had groups_to_compress as 2000, then it would return the 400th, but with OFFSET, the query would return nothing

Reply:
This works: `SELECT id from (SELECT id FROM state_groups WHERE $1 AND id > $2 LIMIT $3) AS ids ORDER BY ids.id desc LIMIT 1;` 

Is this any worse than doing offset?

Reply:
```sql
synapse=# EXPLAIN SELECT id from (SELECT id FROM state_groups WHERE id > 100 LIMIT 1000) AS ids ORDER BY id desc LIMIT 1;
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Limit  (cost=44.78..44.78 rows=1 width=8)
   ->  Sort  (cost=44.78..47.28 rows=1000 width=8)
         Sort Key: state_groups.id DESC
         ->  Limit  (cost=0.43..29.78 rows=1000 width=8)
               ->  Index Only Scan using state_groups_pkey on state_groups  (cost=0.43..66734.45 rows=2273722 width=8)
                     Index Cond: (id > 100)
(6 rows)

```

Looks fine, basically it scans the index starting from the given ID, sorts it, and then takes the last one. It's not quite as efficient as offset in that it accumulates all the ids into a list, but this is a lot better than returning the list to the app :slightly_smiling_face:

Reply:
done